### DISCUSSION TITLE
Grouping a list of similar structs by field value

### DISCUSSION THREAD

**Discussion by zloeber** [cuelang/cue#679] [General]
It can be beneficial to be able to group a list of similar structs by a common field name.  As shown below, Cuelang accomplish this quite easily.

```
// A hidden list definition
_#mylist: [ 
    { 
        animal: "Cat"
        sound: "Meow!"
        type: "mammal"
    }, {
        animal: "Dog"
        type: "mammal"
    }, {
        animal: "Mouse"
        type: "mammal"
    }, {
        animal: "Lizard"
        type: "reptile"
    }
]

// First, create a list of animal types using field interpolation in a new struct 
//   wherein fields are  using the 'type' value as their name (for automatic deduplication).
_#types: [ for type in _#mylist { "\(type.type)" } ]

// Then, using each type in the generated list, group the original list of structs using an `if` guard
// Example of grouping element by type
grouped: {
    for type in _#types {
        "\(type)": [for item in _#mylist if item.type == type { item.animal } ]
    }
}

// Same as above but keeping all original attributes
grouped_full: {
    for type in _#types {
        "\(type)": [for item in _#mylist if item.type == type { item } ]
    }
}
```
See this in action --> https://cuelang.org/play/?id=u0BdnMvW_Cf#cue@export@cue

**Comment by verdverm** [comment]
It is a little confusing to use the word "attribute" because Cue has a separate concept for attributes. Cue does not work with attributes directly (i.e. you cannot compute or unify with them) and they are only accessible via the Go API. Cue does this with `@tag` and injection.

https://cuelang.org/docs/references/spec/#attributes

Maybe 'field' or 'field values' would be better?

**Reply by zloeber** [reply]
Thanks, updated!

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/679.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
